src/vcpkg/dependencies.cpp

```diff
@@ -837,16 +861,41 @@ namespace vcpkg::Dependencies
                                                     spec.name(),
                                                     spec.feature());
                         paragraph_depends = &maybe_paragraph.value_or_exit(VCPKG_LINE_INFO).dependencies;
+                        has_supports = !maybe_paragraph.get()->supports_expression.is_empty();
                     }
 
                     // And it has at least one qualified dependency
-                    if (paragraph_depends &&
-                        Util::any_of(*paragraph_depends, [](auto&& dep) { return !dep.platform.is_empty(); }))
+                    if (has_supports || (paragraph_depends && Util::any_of(*paragraph_depends, [](auto&& dep) {
+                                             return !dep.platform.is_empty();
+                                         })))
                     {
                         // Add it to the next batch run
                         qualified_dependencies.emplace_back(spec);
                     }
                 }
+                else
+                {
+                    auto supports_expression = clust.get_applicable_supports_expression(spec);
+                    if (supports_expression && !supports_expression.get()->is_empty())
+                    {
+                        if (!supports_expression.get()->evaluate(
+                                m_var_provider.get_dep_info_vars(spec.spec()).value_or_exit(VCPKG_LINE_INFO)))
+                        {
+                            const auto msg = Strings::format("%s[%s] is only supported on '%s'",
```

Comment:
I strongly against this changes!
Since our document does not specify whether to use a whitelist or a blacklist for the supports field, this modification will cause many ports to fail to build on the community triplet.
Please note that this field is only used for pipeline test to judge whether the result is reasonable.

Reply:
Support expressions supports black and whitelists. For example `!windows` is a blacklist while `osx & linux` is a whitelist. But yeah we should suggest using blacklists to not fail on community triplets unless we know that a port really only works the given set of triplets. 

>  Please note that this field is only used for pipeline test to judge whether the result is reasonable.

Until this PR. Now it can replaces `vcpkg_fail_port_install` calls to get a better user experience. You don't have to build all dependencies to see that the port does not support your triplet.  
If a port is generally supported by a triplet but our ci don't want to build it, you should use `ci.baseline.txt` because the port in general supports the triplet.

Reply:
>Please note that this field is only used for pipeline test to judge whether the result is reasonable.

The intent is that ci.baseline.txt is the 'specific to our pipelines' thing, while `supports` reflects behavior from the upstream sources.

Reply:
Therefore, I suggest changes this error to warning message.

Reply:
This also breaks option `--only-downloads` and `--keep-going`.

Reply:
You can use the option `--allow-unsupported`

Reply:
This also breaks command `depend-info` which should nevert fails.

Reply:
> This also breaks command `depend-info` which should nevert fails.

I don't think depend-info is broken. It works as expected for the target triplet, listing the unsupported dependency when needed.

Reply:
@dg0yt depend-info is only used to view dependencies without affecting the build, so I don't think we should stop it here.

Reply:
@JackBoosY The current behaviour allows me to inspect unsupported dependencies for arm-uwp before attempting to push to CI from x64-linux. This is beneficial.

Reply:
> @JackBoosY The current behaviour allows me to inspect unsupported dependencies for arm-uwp before attempting to push to CI from x64-linux. This is beneficial.

At lest depend-info doesn't support extra arg `--allow-unsupported`.

Reply:
.. and it doesn't allow multiple package names, to limit the feature set of dependencies. :disappointed:

Reply:
And this doesn't exists in the usage info.

Reply:
> And this doesn't exists in the usage info.

I don't understand that point.

Reply:
> I don't understand that point.

This option should be printed in the usage info if we need to add it:
```
PS F:\vcpkg> .\vcpkg.exe depend-info vtk[*] --allow-unsupported
Unknown option(s) for command 'depend-info':
    '--allow-unsupported'

Example:
  vcpkg depend-info sqlite3

Options:
  --dot                           Creates graph on basis of dot
  --dgml                          Creates graph on basis of dgml
  --show-depth                    Show recursion depth in output
  --max-recurse=...               Set max recursion depth, a value of -1 indicates no limit
  --sort=...                      Set sort order for the list of dependencies, accepted values are:
                                  lexicographical, topological (default), reverse
  --triplet=<t>                   Specify the target architecture triplet. See 'vcpkg help triplet'
                                  (default: %VCPKG_DEFAULT_TRIPLET%)
  --host-triplet=<t>              Specify the host architecture triplet. See 'vcpkg help triplet'
                                  (default: %VCPKG_DEFAULT_HOST_TRIPLET%)
  --overlay-ports=<path>          Specify directories to be used when searching for ports
                                  (also: %VCPKG_OVERLAY_PORTS%)
  --overlay-triplets=<path>       Specify directories containing triplets files
                                  (also: %VCPKG_OVERLAY_TRIPLETS%)
  --binarysource=<path>           Add sources for binary caching. See 'vcpkg help binarycaching'
  --x-asset-sources=<path>        Add sources for asset caching. See 'vcpkg help assetcaching'
  --downloads-root=<path>         Specify the downloads root directory
                                  (default: %VCPKG_DOWNLOADS%)
  --vcpkg-root=<path>             Specify the vcpkg root directory
                                  (default: %VCPKG_ROOT%)
  --x-buildtrees-root=<path>      (Experimental) Specify the buildtrees root directory
  --x-install-root=<path>         (Experimental) Specify the install root directory
  --x-packages-root=<path>        (Experimental) Specify the packages root directory
  --x-json                        (Experimental) Request JSON output
```

Reply:
Fix this in https://github.com/microsoft/vcpkg-tool/pull/300.

Reply:
Another issue: https://github.com/microsoft/vcpkg/issues/22278